### PR TITLE
Coaddition

### PR DESCRIPTION
**Pull request by julienguy**
- Fix bug in coaddition with nsig>0 for cosmic ray rejection (a reference instead of a copy of the flux and ivar were used in the outlier identification routine)/
 - Set nsig=4 to new default
 - Add scripts `desi_coadd_frames` and `desi_coadd_preproc` that are useful for some studies.

I would like to add to the PR the possibility to coadd several spectra files (to be discussed). For now the code `desi_coadd_spectra` accepts either a single spectra file (with internally several exposures) or several frame files.

### CODE REVIEW

**Comment by julienguy** [comment]
This solves issue #817

**Comment by julienguy** [comment]
Added possibility to coadd several spectra files. For instance, 
`desi_coadd_spectra -i $DESI_SPECTRO_REDUX/blanc/tiles/80608/202012*/coadd-0-80608-*.fits -o coadd-0-80608.fits`

File: py/desispec/scripts/coadd_frames.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    mask = frames[0].mask
+    for frame in frames[1:] :
+        ivar += frame.ivar * (frame.mask == 0)
+        ivarflux += frame.ivar * (frame.mask == 0) * frame.flux
+        mask &= frame.mask # and mask
+
+    coadd = frames[0]
+    coadd.ivar = ivar
+    coadd.flux = ivarflux/(ivar+(ivar==0))
+    coadd.mask = mask
```
Comment (sbailey) [inline py/desispec/scripts/coadd_frames.py:71]: What about other `Frame` quantities: fibermap, chi2pix, resolution_data, scores?

I'm concerned about the maintainability of coading spectra vs. coadding frames, which are doing the same basic operations in different files, and possibly making different choices about things like how/if to coadd scores, what to do with the resolution matrix, what to do with masked ivars, etc.

File: py/desispec/scripts/coadd_spectra.py [hunk 2]
```diff
@@ -54,12 +55,46 @@ def main(args=None):
...
+        log.error("cannot combine input spectra and frames")
+        sys.exit(1)
+
+
+    if input_is_spectra :
         spectra = read_spectra(args.infile[0])
-    else:
+        for filename in args.infile[1:] :
+            log.info("append {}".format(filename))
+            spectra.update(read_spectra(filename))
```
Comment (sbailey) [inline py/desispec/scripts/coadd_spectra.py:96]: We should consider merging #1107 and then using `desispec.spectra.stack` instead which is way faster than `Spectra.update`, albeit with the restriction that the input spectra have to be homogeneous (e.g. they all have the same cameras).

**Review by sbailey — COMMENTED** [review]
Thanks for fixing nsig, and adding more coadding flexibility.  Two issues for consideration:
  * it looks like you are coadding some frame quantities (flux, ivar) but not others (resolution matrix, scores).
  * we're building up multiple coadd functions for the same basic operations on frames vs. spectra.  Can this be consolidated for maintainability and consistent choices?

File: py/desispec/scripts/coadd_frames.py (hunk 1 above)
Comment (julienguy) [reply py/desispec/scripts/coadd_frames.py:71]: The use case for this script is debugging the spectral processing. 

For this we probably won't use chi2pix and scores. So I will remove chi2pix and scores, and update resolution_data. I am not sure how to "average" fibermaps. I keep the first one for now and check the target ids are the same.

File: py/desispec/scripts/coadd_spectra.py (hunk 2 above)
Comment (julienguy) [reply py/desispec/scripts/coadd_spectra.py:96]: ok

Comment (sbailey) [reply py/desispec/scripts/coadd_spectra.py:96]: to avoid rebasing confusion between two different open PRs, we can address this optimization later.

**Review by sbailey — APPROVED** [review]

**Pull request merged**